File: app/src/main/java/org/mifos/selfserviceapp/ui/fragments/SavingsMakeTransferFragment.java
package org.mifos.selfserviceapp.ui.fragments;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.widget.AppCompatButton;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.Spinner;
import android.widget.TextView;

import org.mifos.selfserviceapp.R;
import org.mifos.selfserviceapp.models.payload.TransferPayload;
import org.mifos.selfserviceapp.models.templates.account.AccountOption;
import org.mifos.selfserviceapp.models.templates.account.AccountOptionsTemplate;
import org.mifos.selfserviceapp.presenters.SavingsMakeTransferPresenter;
import org.mifos.selfserviceapp.ui.activities.base.BaseActivity;
import org.mifos.selfserviceapp.ui.enums.TransferType;
import org.mifos.selfserviceapp.ui.fragments.base.BaseFragment;
import org.mifos.selfserviceapp.ui.views.SavingsMakeTransferMvpView;
import org.mifos.selfserviceapp.utils.Constants;
import org.mifos.selfserviceapp.utils.DateHelper;
import org.mifos.selfserviceapp.utils.MFDatePicker;
import org.mifos.selfserviceapp.utils.ProcessView;
import org.mifos.selfserviceapp.utils.Toaster;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

/**
 * Created by Rajan Maurya on 10/03/17.
 */
public class SavingsMakeTransferFragment extends BaseFragment implements
        SavingsMakeTransferMvpView, AdapterView.OnItemSelectedListener {

    @BindView(R.id.sp_pay_to)
    Spinner spPayTo;

    @BindView(R.id.sp_pay_from)
    Spinner spPayFrom;

    @BindView(R.id.et_amount)
    EditText etAmount;

    @BindView(R.id.process_one)
    ProcessView pvOne;

    @BindView(R.id.process_two)
    ProcessView pvTwo;

    @BindView(R.id.process_three)
    ProcessView pvThree;

    @BindView(R.id.process_four)
    ProcessView pvFour;

    @BindView(R.id.btn_pay_to)
    AppCompatButton btnPayTo;

    @BindView(R.id.btn_pay_from)
    AppCompatButton btnPayFrom;

    @BindView(R.id.btn_amount)
    AppCompatButton btnAmount;

    @BindView(R.id.ll_review)
    LinearLayout llReview;

    @BindView(R.id.tv_select_pay_from)
    TextView tvSelectPayFrom;

    @BindView(R.id.tv_select_amount)
    TextView tvEnterAmount;

    @BindView(R.id.tv_enter_remark)
    TextView tvEnterRemark;

    @BindView(R.id.et_remark)
    EditText etRemark;

    @BindView(R.id.ll_make_transfer)
    LinearLayout layoutMakeTransfer;

    @Inject
    SavingsMakeTransferPresenter savingsMakeTransferPresenter;

    View rootView;

    private List<String> listPayTo = new ArrayList<>();
    private List<String> listPayFrom = new ArrayList<>();

    private ArrayAdapter<String> payToAdapter;
    private ArrayAdapter<String> payFromAdapter;

    private TransferPayload transferPayload;
    private String transferDate;
    private AccountOption toAccountOption, fromAccountOption;
    private AccountOptionsTemplate accountOptionsTemplate;
    private String transferType;
    private long accountId;

    public static SavingsMakeTransferFragment newInstance(long accountId, String transferType) {
        SavingsMakeTransferFragment transferFragment = new SavingsMakeTransferFragment();
        Bundle args = new Bundle();
        args.putLong(Constants.ACCOUNT_ID, accountId);
        args.putString(Constants.TRANSFER_TYPE, transferType);
        transferFragment.setArguments(args);
        return transferFragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            accountId = getArguments().getLong(Constants.ACCOUNT_ID);
            transferType = getArguments().getString(Constants.TRANSFER_TYPE);
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
            @Nullable Bundle savedInstanceState) {
        rootView = inflater.inflate(R.layout.fragment_savings_make_transfer, container, false);
        ((BaseActivity) getActivity()).getActivityComponent().inject(this);
        setToolbarTitle(getString(R.string.transfer));
        ButterKnife.bind(this, rootView);
        savingsMakeTransferPresenter.attachView(this);

        showUserInterface();
        savingsMakeTransferPresenter.loanAccountTransferTemplate();

        return rootView;
    }

    @OnClick(R.id.btn_review_transfer)
    void reviewTransfer() {

        if (etRemark.getText().toString().equals("")) {
            showToaster(getString(R.string.remark_is_mandatory));
            return;
        }

        transferPayload = new TransferPayload();
        transferPayload.setFromAccountId(fromAccountOption.getAccountId());
        transferPayload.setFromClientId(fromAccountOption.getClientId());
        transferPayload.setFromAccountType(fromAccountOption.getAccountType().getId());
        transferPayload.setFromOfficeId(fromAccountOption.getOfficeId());
        transferPayload.setToOfficeId(toAccountOption.getOfficeId());
        transferPayload.setToAccountId(toAccountOption.getAccountId());
        transferPayload.setToClientId(toAccountOption.getClientId());
        transferPayload.setToAccountType(toAccountOption.getAccountType().getId());
        transferPayload.setTransferDate(transferDate);
        transferPayload.setTransferAmount(Double.parseDouble(etAmount.getText().toString()));
        transferPayload.setTransferDescription(etRemark.getText().toString());


        ((BaseActivity) getActivity()).replaceFragment(TransferProcessFragment.
                newInstance(transferPayload, TransferType.SELF), true, R.id.container);
    }

    @OnClick(R.id.btn_cancel_transfer)
    void cancelTransfer() {
        getActivity().getSupportFragmentManager().popBackStack();
    }

    @Override
    public void showUserInterface() {
        pvOne.setCurrentActive();
        payFromAdapter = new ArrayAdapter<>(getActivity(), android.R.layout.simple_spinner_item,
                listPayFrom);
        payFromAdapter.setDropDownViewResource(android.R.layout.select_dialog_singlechoice);
        spPayFrom.setAdapter(payFromAdapter);
        spPayFrom.setOnItemSelectedListener(this);

        payToAdapter = new ArrayAdapter<>(getActivity(), android.R.layout.simple_spinner_item,
                listPayTo);
        payToAdapter.setDropDownViewResource(android.R.layout.select_dialog_singlechoice);
        spPayTo.setAdapter(payToAdapter);
        spPayTo.setOnItemSelectedListener(this);
        transferDate = DateHelper.getSpecificFormat(DateHelper.FORMAT_dd_MMMM_yyyy,
                MFDatePicker.getDatePickedAsString());
    }

    @Override
    public void showSavingsAccountTemplate(AccountOptionsTemplate accountOptionsTemplate) {
        this.accountOptionsTemplate = accountOptionsTemplate;
        listPayFrom.addAll(savingsMakeTransferPresenter.getAccountNumbers(
                accountOptionsTemplate.getFromAccountOptions()));
        listPayTo.addAll(savingsMakeTransferPresenter.getAccountNumbers(
                accountOptionsTemplate.getToAccountOptions()));
        payToAdapter.notifyDataSetChanged();
        payFromAdapter.notifyDataSetChanged();
    }

    @Override
    public void showToaster(String message) {
        Toaster.show(rootView, message);
    }

    @Override
    public void showError(String message) {
        Toaster.show(rootView, message);
    }

    @Override
    public void showProgressDialog() {
        showMifosProgressDialog(getString(R.string.making_transfer));
    }

    @Override
    public void hideProgressDialog() {
        hideMifosProgressDialog();
    }

    @Override
    public void showProgress() {
        layoutMakeTransfer.setVisibility(View.GONE);
        showProgressBar();
    }

    @Override
    public void hideProgress() {
        layoutMakeTransfer.setVisibility(View.VISIBLE);
        hideProgressBar();
    }

    @Override
    public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
        switch (parent.getId()) {
            case R.id.sp_pay_to:
                toAccountOption = accountOptionsTemplate.getToAccountOptions().get(position);
                break;
            case R.id.sp_pay_from:
                fromAccountOption = accountOptionsTemplate.getFromAccountOptions().get(position);
                break;
        }

        switch (transferType) {
            case Constants.TRANSFER_PAY_TO:
                setToolbarTitle(getString(R.string.deposit));
                toAccountOption = savingsMakeTransferPresenter
                        .searchAccount(accountOptionsTemplate.getToAccountOptions(), accountId);
                spPayTo.setSelection(accountOptionsTemplate.getToAccountOptions()
                        .indexOf(toAccountOption));
                spPayTo.setEnabled(false);
                pvOne.setCurrentCompeleted();
                break;
            case Constants.TRANSFER_PAY_FROM:
                setToolbarTitle(getString(R.string.transfer));
                fromAccountOption = savingsMakeTransferPresenter
                        .searchAccount(accountOptionsTemplate.getFromAccountOptions(), accountId);
                spPayFrom.setSelection(accountOptionsTemplate.getFromAccountOptions()
                        .indexOf(fromAccountOption));
                spPayFrom.setEnabled(false);
                spPayFrom.setVisibility(View.VISIBLE);
                tvSelectPayFrom.setVisibility(View.GONE);
                pvTwo.setCurrentCompeleted();
                break;
        }
    }

    @Override
    public void onNothingSelected(AdapterView<?> parent) {

    }

    @OnClick(R.id.btn_pay_to)
    public void payToSelected() {
        pvOne.setCurrentCompeleted();
        pvTwo.setCurrentActive();

        btnPayTo.setVisibility(View.GONE);
        tvSelectPayFrom.setVisibility(View.GONE);
        btnPayFrom.setVisibility(View.VISIBLE);
        spPayFrom.setVisibility(View.VISIBLE);
        spPayTo.setEnabled(false);
    }

    @OnClick(R.id.btn_pay_from)
    public void payFromSelected() {
        if (spPayTo.getSelectedItem().toString().equals(spPayFrom.getSelectedItem().toString())) {
            showToaster(getString(R.string.error_same_account_transfer));
            return;
        }
        pvTwo.setCurrentCompeleted();
        pvThree.setCurrentActive();

        btnPayFrom.setVisibility(View.GONE);
        tvEnterAmount.setVisibility(View.GONE);
        etAmount.setVisibility(View.VISIBLE);
        btnAmount.setVisibility(View.VISIBLE);
        spPayFrom.setEnabled(false);
    }

    @OnClick(R.id.btn_amount)
    public void amountSet() {

        if (etAmount.getText().toString().equals("")) {
            showToaster(getString(R.string.enter_amount));
            return;
        }

        if (etAmount.getText().toString().equals(".")) {
            showToaster(getString(R.string.invalid_amount));
            return;
        }

        pvThree.setCurrentCompeleted();
        pvFour.setCurrentActive();

        btnAmount.setVisibility(View.GONE);
        tvEnterRemark.setVisibility(View.GONE);
        etRemark.setVisibility(View.VISIBLE);
        llReview.setVisibility(View.VISIBLE);
        etAmount.setEnabled(false);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        hideProgress();
        hideMifosProgressDialog();
        savingsMakeTransferPresenter.detachView();
    }
}

File: app/src/main/java/org/mifos/mobile/ui/savings_make_transfer/SavingsMakeTransferFragment.kt
package org.mifos.mobile.ui.savings_make_transfer

import dagger.hilt.android.AndroidEntryPoint
import org.mifos.mobile.ui.fragments.base.BaseFragment

/**
 * Created by Rajan Maurya on 10/03/17.
 */
@AndroidEntryPoint
class SavingsMakeTransferFragment : BaseFragment() {
//
//    private var _binding: FragmentSavingsMakeTransferBinding? = null
//    private val binding get() = _binding!!
//    private val viewModel: SavingsMakeTransferViewModel by viewModels()
//    private var transferPayload: TransferPayload? = null
//    private var transferDate: String? = null
//    private var toAccountOption: AccountOption? = null
//    private var fromAccountOption: AccountOption? = null
//    private var accountOptionsTemplate: AccountOptionsTemplate? = null
//    private var transferType: String? = null
//    private var payTo: String? = null
//    private var payFrom: String? = null
//    private var accountId: Long? = 0
//    private var outStandingBalance: Double? = 0.0
//    private var isLoanRepayment = false
//    private var sweetUIErrorHandler: SweetUIErrorHandler? = null
//
//    override fun onCreate(savedInstanceState: Bundle?) {
//        super.onCreate(savedInstanceState)
//        (activity as? BaseActivity)?.showToolbar()
//        if (arguments != null) {
//            accountId = arguments?.getLong(Constants.ACCOUNT_ID)
//            transferType = arguments?.getString(Constants.TRANSFER_TYPE)
//            if (arguments?.getBoolean(Constants.LOAN_REPAYMENT, false) == true) {
//                isLoanRepayment = true
//                outStandingBalance = arguments?.getDouble(Constants.OUTSTANDING_BALANCE)
//            }
//        }
//        setHasOptionsMenu(true)
//    }
//
//    override fun onCreateView(
//        inflater: LayoutInflater,
//        container: ViewGroup?,
//        savedInstanceState: Bundle?,
//    ): View {
//        _binding = FragmentSavingsMakeTransferBinding.inflate(inflater, container, false)
//        setToolbarTitle(getString(R.string.transfer))
//        sweetUIErrorHandler = SweetUIErrorHandler(activity, binding.root)
//        showUserInterface()
//        if (savedInstanceState == null) {
//            //viewModel.loanAccountTransferTemplate()
//        }
//        return binding.root
//    }
//
//    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
//        super.onViewCreated(view, savedInstanceState)
//
//        with(binding) {
//            btnCancelTransfer.setOnClickListener {
//                cancelTransfer()
//            }
//
//            btnReviewTransfer.setOnClickListener {
//                reviewTransfer()
//            }
//
//            btnPayFrom.setOnClickListener {
//                payFromSelected()
//            }
//
//            btnPayTo.setOnClickListener {
//                payToSelected()
//            }
//
//            btnAmount.setOnClickListener {
//                amountSet()
//            }
//
//            layoutError.btnTryAgain.setOnClickListener {
//                onRetry()
//            }
//        }
//
////        viewLifecycleOwner.lifecycleScope.launch {
////            repeatOnLifecycle(Lifecycle.State.STARTED) {
////                viewModel.savingsMakeTransferUiState.collect { state ->
////                    when (state) {
////                        SavingsAccountUiState.Loading -> showProgress()
////
////                        is SavingsAccountUiState.ErrorMessage -> {
////                            hideProgress()
////                            showError(context?.getString(R.string.error_fetching_account_transfer_template))
////                        }
////
////                        is SavingsAccountUiState.ShowSavingsAccountTemplate -> {
////                            hideProgress()
////                            showSavingsAccountTemplate(state.accountOptionsTemplate)
////                        }
////
////                        is SavingsAccountUiState.Initial -> {}
////
////                        else -> throw IllegalStateException("Unexpected state : $state")
////                    }
////                }
////            }
////        }
//    }
//
//    override fun onSaveInstanceState(outState: Bundle) {
//        super.onSaveInstanceState(outState)
//        outState.putParcelable(Constants.TEMPLATE, accountOptionsTemplate)
//    }
//
//    override fun onActivityCreated(savedInstanceState: Bundle?) {
//        super.onActivityCreated(savedInstanceState)
//        if (savedInstanceState != null) {
//            showSavingsAccountTemplate(
//                savedInstanceState.getCheckedParcelable(
//                    AccountOptionsTemplate::class.java,
//                    Constants.TEMPLATE
//                )
//            )        }
//    }
//
//    /**
//     * Checks validation of `etRemark` and then opens [TransferProcessComposeFragment] for
//     * initiating the transfer
//     */
//    private fun reviewTransfer() {
//        if (binding.etRemark.text.toString().trim { it <= ' ' } == "") {
//            showToaster(getString(R.string.remark_is_mandatory))
//            return
//        }
//        transferPayload = TransferPayload()
//        transferPayload?.fromAccountId = fromAccountOption?.accountId
//        transferPayload?.fromClientId = fromAccountOption?.clientId
//        transferPayload?.fromAccountType = fromAccountOption?.accountType?.id
//        transferPayload?.fromOfficeId = fromAccountOption?.officeId
//        transferPayload?.toOfficeId = toAccountOption?.officeId
//        transferPayload?.toAccountId = toAccountOption?.accountId
//        transferPayload?.toClientId = toAccountOption?.clientId
//        transferPayload?.toAccountType = toAccountOption?.accountType?.id
//        transferPayload?.transferDate = transferDate
//        transferPayload?.transferAmount = binding.amountField.text.toString().toDouble()
//        transferPayload?.transferDescription = binding.etRemark.text.toString()
//        transferPayload?.fromAccountNumber = fromAccountOption?.accountNo
//        transferPayload?.toAccountNumber = toAccountOption?.accountNo
//        (activity as BaseActivity?)?.replaceFragment(
//            TransferProcessComposeFragment.newInstance(
//                transferPayload,
//                TransferType.SELF,
//            ),
//            true,
//            R.id.container,
//        )
//    }
//
//    /**
//     * Cancels the transfer by popping current Fragment
//     */
//    private fun cancelTransfer() {
//        activity?.supportFragmentManager?.popBackStack()
//    }
//
//    private fun onRetry() {
//        if (Network.isConnected(context)) {
//            sweetUIErrorHandler?.hideSweetErrorLayoutUI(
//                binding.llMakeTransfer,
//                binding.layoutError.root,
//            )
//            //viewModel.loanAccountTransferTemplate()
//        } else {
//            Toaster.show(binding.root, getString(R.string.internet_not_connected))
//        }
//    }
//
//    /**
//     * Setting up basic components
//     */
//    fun showUserInterface() {
//        binding.processOne.setCurrentActive()
//        binding.payFromField.setOnItemClickListener { _, _, position, _ ->
//            toAccountOption = accountOptionsTemplate?.toAccountOptions?.get(position)
//            payTo = toAccountOption?.accountNo
//            println("paytofield item selected payTo:$payTo")
//            updateDetails()
//        }
//        binding.payToField.setOnItemClickListener { _, _, position, _ ->
//            fromAccountOption = accountOptionsTemplate?.fromAccountOptions?.get(position)
//            payFrom = fromAccountOption?.accountNo
//            println("payfrom item selected payFrom:$payFrom")
//            updateDetails()
//        }
//        transferDate = DateHelper.getSpecificFormat(
//            DateHelper.FORMAT_dd_MMMM_yyyy,
//            getTodayFormatted(),
//        )
//        if (isLoanRepayment) {
//            binding.amountField.setText(outStandingBalance.toString())
//            binding.amountFieldWrapper.isFocusable = false
//        }
//    }
//
//    private fun updateDetails() {
////        when (transferType) {
////            Constants.TRANSFER_PAY_TO -> {
////                setToolbarTitle(getString(R.string.deposit))
////                toAccountOption =
////                    viewModel.searchAccount(accountOptionsTemplate?.toAccountOptions, accountId)
////                binding.payToFieldWrapper.isEnabled = false
////                binding.processOne.setCurrentCompleted()
////            }
////
////            Constants.TRANSFER_PAY_FROM -> {
////                setToolbarTitle(getString(R.string.transfer))
////                fromAccountOption =
////                    viewModel.searchAccount(accountOptionsTemplate?.fromAccountOptions, accountId)
////                binding.payFromFieldWrapper.isEnabled = false
////                binding.payFromFieldWrapper.visibility = View.VISIBLE
////                binding.processTwo.setCurrentCompleted()
////            }
////        }
//    }
//
//    /**
//     * Provides with `accountOptionsTemplate` fetched from server which is used to update
//     * `listPayFrom` and `listPayTo`
//     *
//     * @param accountOptionsTemplate Template for account transfer
//     */
//    private fun showSavingsAccountTemplate(accountOptionsTemplate: AccountOptionsTemplate?) {
//        this.accountOptionsTemplate = accountOptionsTemplate
////        binding.payToField.setAdapter(
////            AccountsSpinnerAdapter(
////                requireContext(),
////                viewModel.getAccountNumbers(
////                    accountOptionsTemplate?.toAccountOptions,
////                    false,
////                    context?.getString(R.string.account_type_loan)
////                ),
////            ),
////        )
////        binding.payFromField.setAdapter(
////            AccountsSpinnerAdapter(
////                requireContext(),
////                viewModel.getAccountNumbers(
////                    accountOptionsTemplate?.toAccountOptions,
////                    true,
////                    context?.getString(R.string.account_type_loan)
////                ),
////            ),
////        )
//    }
//
//    /**
//     * Shows a {@link Snackbar} with `message`
//     *f
//     * @param message String to be shown
//     */
//    private fun showToaster(message: String?) {
//        Toaster.show(binding.root, message)
//    }
//
//    /**
//     * It is called whenever any error occurs while executing a request
//     *
//     * @param message Error message that tells the user about the problem.
//     */
//    fun showError(message: String?) {
//        if (!Network.isConnected(context)) {
//            sweetUIErrorHandler?.showSweetNoInternetUI(
//                binding.llMakeTransfer,
//                binding.layoutError.root,
//            )
//        } else {
//            sweetUIErrorHandler?.showSweetErrorUI(
//                message,
//                binding.llMakeTransfer,
//                binding.layoutError.root,
//            )
//            Toaster.show(binding.root, message)
//        }
//    }
//
//    fun showProgress() {
//        binding.llMakeTransfer.visibility = View.GONE
//        showProgressBar()
//    }
//
//    fun hideProgress() {
//        binding.llMakeTransfer.visibility = View.VISIBLE
//        hideProgressBar()
//    }
//
//    /**
//     * Callback for `spPayFrom` and `spPayTo`
//     */
//
//    /**
//     * Disables `spPayTo` [Spinner] and sets `pvOne` to completed and make
//     * `pvTwo` active
//     */
//    private fun payToSelected() {
//        with(binding) {
//            processOne.setCurrentCompleted()
//            processTwo.setCurrentActive()
//            btnPayTo.visibility = View.GONE
//            btnPayFrom.visibility = View.VISIBLE
//            payFromFieldWrapper.visibility = View.VISIBLE
//            payToFieldWrapper.isEnabled = false
//        }
//    }
//
//    /**
//     * Checks validation of `spPayTo` [Spinner].<br></br>
//     * Disables `spPayFrom` [Spinner] and sets `pvTwo` to completed and make
//     * `pvThree` active
//     */
//    private fun payFromSelected() {
//        if (payTo == payFrom) {
//            showToaster(getString(R.string.error_same_account_transfer))
//            return
//        }
//        with(binding) {
//            processTwo.setCurrentCompleted()
//            processThree.setCurrentActive()
//            btnPayFrom.visibility = View.GONE
//            amountFieldWrapper.visibility = View.VISIBLE
//            btnAmount.visibility = View.VISIBLE
//            payFromFieldWrapper.isEnabled = false
//        }
//    }
//
//    /**
//     * Checks validation of `etAmount` [EditText].<br></br>
//     * Disables `etAmount` and sets `pvThree` to completed and make
//     * `pvFour` active
//     */
//    private fun amountSet() {
//        if (binding.amountField.text.toString() == "") {
//            showToaster(getString(R.string.enter_amount))
//            return
//        }
//        if (binding.amountField.text.toString() == ".") {
//            showToaster(getString(R.string.invalid_amount))
//            return
//        }
//        if (binding.amountField.text.toString().toDouble() == 0.0) {
//            showToaster(getString(R.string.amount_greater_than_zero))
//            return
//        }
//        with(binding) {
//            processThree.setCurrentCompleted()
//            processFour.setCurrentActive()
//            btnAmount.visibility = View.GONE
//            tvEnterRemark.visibility = View.GONE
//            etRemark.visibility = View.VISIBLE
//            llReview.visibility = View.VISIBLE
//            amountFieldWrapper.isEnabled = false
//        }
//    }
//
//    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
//        super.onCreateOptionsMenu(menu, inflater)
//        inflater.inflate(R.menu.menu_transfer, menu)
//        Utils.setToolbarIconColor(activity, menu, R.color.white)
//    }
//
//    override fun onOptionsItemSelected(item: MenuItem): Boolean {
//        if (item.itemId == R.id.menu_refresh_transfer) {
//            val transaction = fragmentManager?.beginTransaction()
//            val currFragment = activity?.supportFragmentManager
//                ?.findFragmentById(R.id.container)
//            if (currFragment != null) {
//                transaction?.detach(currFragment)
//                transaction?.attach(currFragment)
//            }
//            transaction?.commit()
//        }
//        return super.onOptionsItemSelected(item)
//    }
//
//    override fun onDestroyView() {
//        super.onDestroyView()
//        hideProgress()
//        hideMifosProgressDialog()
//        _binding = null
//    }
//
//    companion object {
//        /**
//         * Provides an instance of [SavingsMakeTransferFragment], use `transferType` as
//         * `Constants.TRANSFER_PAY_TO` when we want to deposit and
//         * `Constants.TRANSFER_PAY_FROM` when we want to make a transfer
//         *
//         * @param accountId    Saving account Id
//         * @param transferType Type of transfer i.e. `Constants.TRANSFER_PAY_TO` or
//         * `Constants.TRANSFER_PAY_FROM`
//         * @return Instance of [SavingsMakeTransferFragment]
//         */
//        fun newInstance(accountId: Long?, transferType: String?): SavingsMakeTransferFragment {
//            val transferFragment = SavingsMakeTransferFragment()
//            val args = Bundle()
//            if (accountId != null) args.putLong(Constants.ACCOUNT_ID, accountId)
//            args.putString(Constants.TRANSFER_TYPE, transferType)
//            transferFragment.arguments = args
//            return transferFragment
//        }
//
//        fun newInstance(
//            accountId: Long?,
//            outstandingBalance: Double?,
//            transferType: String?,
//        ): SavingsMakeTransferFragment {
//            val transferFragment = SavingsMakeTransferFragment()
//            val args = Bundle()
//            if (accountId != null) args.putLong(Constants.ACCOUNT_ID, accountId)
//            args.putString(Constants.TRANSFER_TYPE, transferType)
//            if (outstandingBalance != null) {
//                args.putDouble(
//                    Constants.OUTSTANDING_BALANCE,
//                    outstandingBalance,
//                )
//            }
//            args.putBoolean(Constants.LOAN_REPAYMENT, true)
//            transferFragment.arguments = args
//            return transferFragment
//        }
//    }
}
